### PR TITLE
Training

### PR DESCRIPTION
**Pull request by azamatsab**
Кажется, ушел совсем не туда. Pytorch lighting не использовал (
LabelSmoothing не поддерживается (

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+torch
```
Comment (exotol) [inline requirements.txt]: requirements.txt нужно перенести в проект, у каждого проекта он свой, по сути каждая из частей - это отдельные проекты. В идеале, тут заводится групповой репозиторий, но т.к. гитхаб не дает такой возможности из коробки,
то приходится помещать все проекты в 1 репозиторий

project/
```suggestion
torch
```
    src/
        <code.py>
    tests/
        test_
    params/
       config's...
    tasks/
       sh задачи
    requirements.txt
   .env - переменные окружения, если нужны
    README
    остальные файлы

Считаем, что любой запуск идет из корня

File: training/models/base_model.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            )
+        return scheduler
+
+    def fit(self, dataset):
+        num_workers = self.config["loader"]["num_workers"]
+        batch_size = self.config["net"]["batch_size"]
+        epochs = self.config["net"]["epoch"]
+        lr = self.config["net"]["lr"]
+
+        if self.config["net"]["optimizer"] == "AdamW":
```
Comment (exotol) [inline training/models/base_model.py]: Таким подходом ты ограничил модели в использовании оптимайзеров
Предлагаю все же сделать как изначально планировалось - модель сама должна знать какой ей оптимайзер нужен, ведь Adam далеко не единственный оптимайзер.
+ лучше не делать через if-else, использовать паттерн стратегия
Опирайся на пример из [pytorch-lightning](https://github.com/PyTorchLightning/pytorch-lightning)

File: training/models/base_model.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    get_cosine_schedule_with_warmup,
+    get_cosine_with_hard_restarts_schedule_with_warmup,
+    get_linear_schedule_with_warmup,
+    get_polynomial_decay_schedule_with_warmup,
+)
+
+
+DIRNAME = Path(__file__).parents[1].resolve() / 'weights'
+
+conf_to_scheduler = {
```
Comment (exotol) [inline training/models/base_model.py]: Это лучше сделать, отдав модели, чтобы не ограничиваться себя лишь этими scheduler'ами,
т.е. нужно добавить в базовый класс модели еще 1 метод и требовать его переопределять

File: training/models/base_model.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    get_constant_schedule,
+    get_constant_schedule_with_warmup,
+    get_cosine_schedule_with_warmup,
+    get_cosine_with_hard_restarts_schedule_with_warmup,
+    get_linear_schedule_with_warmup,
+    get_polynomial_decay_schedule_with_warmup,
+)
+
+
+DIRNAME = Path(__file__).parents[1].resolve() / 'weights'
```
Comment (exotol) [inline training/models/base_model.py]: Пути нужно вынести в конфигурационные файлы, 
чтобы я мог указать потом любой путь

File: training/test/model_tests.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+import os
```
Comment (exotol) [inline training/test/model_tests.py]: По-моему, pycharm, например, чувствителен к имени папки с тестами, и предпочитает tests
Все же предлагаю сделать tests, на случай расширения

File: training/models/base_model.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    "linear": get_linear_schedule_with_warmup,
+    "cosine": get_cosine_schedule_with_warmup,
+    "cosine_w_restarts": get_cosine_with_hard_restarts_schedule_with_warmup,
+    "polynomial": get_polynomial_decay_schedule_with_warmup,
+    "constant": get_constant_schedule,
+    "constant_w_warmup": get_constant_schedule_with_warmup,
+    }
+
+
+class BaseModel:
```
Comment (exotol) [inline training/models/base_model.py]: BaseModel - определяет Серега,
тебе нужно его BaseModel дополнять и править 
А он у тебя одиноко валяется в 
training/pipeline.py

File: training/models/marian_model.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from transformers import MarianMTModel, MarianTokenizer
```
Comment (exotol) [inline training/models/marian_model.py]: Модели должны единообразно определяться по аналогии с модулем translator, в этом модуле чисто тренировка,
и никаких моделей быть не может, для своих целей определи свою модель (для тестовых целей, если полностью рабочая модель нужна, то определи ее в translator)

File: training/README.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+###### Для запуска обучения выполните скрипт training/pipeline.py
```
Comment (exotol) [inline training/README.md:1]: не хватает примеров полноценного запуска

File: training/models/base_model.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        torch.save(self.model.state_dict(), filepath)
+
+
+    def load_model(self, filepath: str):
+        """ Load the trained model
+        """
+        self.model.load_state_dict(torch.load(filepath))
+
+if __name__ == "__main__":
+    MODEL_NAME = 'Helsinki-NLP/opus-mt-en-ru'
```
Comment (exotol) [inline training/models/base_model.py]: Эти все константы должны задаваться в конфиге, чтобы мы не правили код каждый раз

File: training/models/utils.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        smooth_loss = smooth_loss.squeeze(-1)
+
+    nll_loss = nll_loss.sum()  # mean()? Scared to break other math.
+    smooth_loss = smooth_loss.sum()
+    eps_i = epsilon / lprobs.size(-1)
+    loss = (1.0 - epsilon) * nll_loss + eps_i * smooth_loss
+    return loss, nll_loss
+
+
+def calculate_bleu(output_lns, refs_lns, **kwargs) -> dict:
```
Comment (exotol) [inline training/models/utils.py]: вынеси в модуль metrics, т.к. будет несколько метрик и мешать их с loss не хочется

**Review by exotol — COMMENTED** [review]
1 не хватает конфага лога и собственно логирования всех этапов в лог еще (mlflow логирует определенные вещи)
2 твой pipeline - это example? test? или что?
3 BaseModel - абстрактный класс, ты не должен его создавать

> Predictor – внутри работает с некоторым **абстрактным классом BaseModel,** конкретная реализация которого задается в настройках

Нужно наследоваться от BaseModel, чтобы общие методы были определены, остальное допределять/переопределять
